### PR TITLE
AutoHeuristic: Support ranking/pruning choices

### PR DESCRIPTION
**Pull request by AlnisM**
This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133611
* #133608
* #131714
* #131710
* __->__ #131705

### CODE REVIEW

File: torchgen/_autoheuristic/train_decision.py [hunk 1]
```diff
@@ -102,8 +103,19 @@ def predict(self, model, df, feature_columns):
         leaf_ids = model.apply(df[feature_columns])
         return predictions, proba, leaf_ids
 
+    def ranking_num_choices(self):
+        # if the heuristic is used for ranking, this function returns the number
+        # of choices that the heuristic will return
+        return 5
```
Comment (eellison) [inline torchgen/_autoheuristic/train_decision.py:109]: I'm surprised this is hardcoded. Is this not configurable at all ?

**Review by eellison — COMMENTED** [review]
just one question about ranking num choices being hardcoded, also tests could b good

File: torchgen/_autoheuristic/train_decision.py (hunk 1 above)
Comment (AlnisM) [reply torchgen/_autoheuristic/train_decision.py:109]: You can configure it by overriding the method https://github.com/pytorch/pytorch/pull/131714/files#diff-45bd9b9ea64020450836e3496f0da071f133e1065dc62fbbb141bd6f52d6d2b0R69. But I'll make this a command line option instead.

File: torchgen/_autoheuristic/ah_tree.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
+# mypy: ignore-errors
```
[the remark this answers is not in the document]
Comment (AlnisM) [reply torchgen/_autoheuristic/ah_tree.py]: added some docstrings. I think some typehints could be better but I need to configure mypy to work with numpy types. happy to do this in a follow-up PR

**Review by eellison — APPROVED** [review]

**Pull request closed without merging**